Dependencies (rebased) (#67)
adoptingerlang/adoptingerlang: adoptingerlang.org

- Rebar3 defines its own working area within the =_build= directory; it expects to have control over what goes in there, and for you not to track that directory in source control nor rely on its internal structure. When it builds a thing you want to use, it usually outputs the path in the terminal. For example =rebar3 escriptize= will print the path to the generated escript in =_build/default/bin/= after each run.
- All projects are built locally within their own directory based on the Erlang runtime currently loaded in your environment
- All dependencies are fetched to the project's =_build= directory
- All dependencies can define their own dependencies, and rebar3 will recognize this and put them within your own project's =_build= directory as well

> - Since version numbers are unreliable (even with semantic versioning), we consider versions are information for humans, not build tools. In case of libraries that end up declared multiple times, the one closest to the project root is chosen to win ([[https://www.rebar3.org/docs/dependencies][warnings]] are emitted during the first build to let the user know about this), with the assumption that libraries declared closer to the root project are used more thoroughly (and in the case of semver issues, the root project has to pick a winner anyway, which is possible here)

Reviewer: ```suggestion
- Since version numbers are unreliable (even with semantic versioning), we consider versions as information for humans, not build tools. In case of libraries that end up declared multiple times, the one closest to the project root is chosen, while [[https://www.rebar3.org/docs/dependencies][warnings]] are emitted during the first build (it knows it is the "first" because there is no lock for the dependency) to let the user know versions not being used, with the assumption that libraries declared closer to the root project are used more thoroughly
```

I don't understand the last part about semver, maybe just needs rewording.

- Circular dependencies are forbidden

> - Rebar3 supports composable per-project [[https://www.rebar3.org/docs/profiles][_profiles_] that let you segment or combine configuration settings, including which dependencies are to be used only for tests, or in specific environment (say on a specific target OS). There are no restrictions on the number or names of profiles, but there are three profiles, =test=, =docs=, and =global=, which are used by rebar3 automatically for specific tasks.

Reviewer: ```suggestion
- Rebar3 supports composable per-project [[https://www.rebar3.org/docs/profiles][_profiles_] that let you segment or combine configuration settings, such as dependencies that are to only be used for tests or only in a specific environment (say on a specific target OS). There are no restrictions on the number or names of profiles, but there are four profiles, =default=, =test=, =docs=, and =global=, which are used by rebar3 automatically for specific tasks.
```

> - Builds are meant to be repeatable. Dependencies are locked, and only upgraded when specifically asked to. The =rebar3.lock= file will overtake whatever versions are declared in =rebar.config= until =rebar3 upgrade= is called to change them.

Reviewer: ```suggestion
- Builds are meant to be repeatable. Dependencies are locked, and only upgraded when specifically asked to. The version in =rebar3.lock= takes precedence of whatever version is declared in =rebar.config= until =rebar3 upgrade= is called to update the lock file. If the lock file changes but dependencies have already been fetched, for example if the current branch checked out from git is changed, =rebar3 compile= will verify the fetched dependencies are those that the lock file specifies and fetch the locked version if needed.
```

> - Dependencies are always built with the =prod= profile applied, and once built will not be recompiled unless they are upgraded, even though the tool will always make sure they are there, match the lock file, and are compiled. Here "compiled" means that there is a =.app= file and for each module listed in the app file a corresponding beam file is found, but it does _not_ check if there are changes to the source file for that beam module.

Reviewer: ```suggestion
- Dependencies are always built with the =prod= profile applied, and once built will not be recompiled, but the tool will always verify they are there, match the lock file, and are compiled. Here "compiled" means that there is a =.app= file and for each module listed in the app file a corresponding beam file is found, but it does _not_ check if there are changes to the source file for that beam module. If a beam file for a module found in the =.app= file is not found it will compile the source for that beam file found in the dependency's =src=.
```

Author: I think that's probably too complex for just an intro of "here are the basics of how rebar3 works". It's accurate, but it's probably too much,

Reviewer: Yea, but then I'd remove most of the original as well. Take it down to just that it builds in the `prod` profile and that it "verifies" the dependencies each time, which does not include scanning for modified source modules.

- The tool assumes you're developing with a source control mechanism such as =git= or =hg= (mercurial), which means that switching branches may switch dependency versions in the lock file. Since rebar3 verifies dependencies before each build, it will automatically re-fetch libraries to get the locked version for the current branch if there is a change when switching branches.
- To ease contribution and publishing, rebar3 does not natively support ways to use relative paths to declare libraries, since this could make builds brittle, non-repeatable, and non-portable when publishing code
- Knowing relative paths are very common when making changes within the dependencies of a project, [[https://www.rebar3.org/docs/dependencies#section-checkout-dependencies][=_checkouts=]] allow for an automatic way to temporary override of a dependency with a local copy. For other use cases [[https://www.rebar3.org/docs/using-available-plugins#section-vendoring-dependencies][plugins]] allow the creation of [[https://www.rebar3.org/docs/custom-dep-resources][custom resource types]].

> - Rebar3 is not an installer of end-user applications and does not support anything related to that; it aims to generate build artifacts that you can then install through the correct dedicated channels. Rebar3 does not expect or need to ever be on a production device or server.

Reviewer: ```suggestion
- Rebar3 is not an installer or runner of end-user applications and does not support anything related to that; it aims to generate build artifacts that you can then install through the correct dedicated channels. Rebar3 does not expect or need to ever be on a production device or server.
```

> - Rebar3 is not a sandbox. While it will make sure all dependencies you download match the right signatures and provides repeatable builds, it cannot guarantee that script files invoked during builds or [[https://stackoverflow.com/questions/2416192/is-there-a-good-complete-tutorial-on-erlang-parse-transforms-available][parse transforms]] run during compilation are ever going to be safe, and has no intention of taking on that responsibility.

Reviewer: May want to mention there are caveats to "repeatable builds" here, or maybe just say "attempts to provide repeatable builds" since this isn't the place to detail issues like plugins not being locked and shit like that.

> That's a lot of information, but we find that it's useful to know about it before getting in too deep with dependencies. If you operate under assumptions that Rebar3 works like Javascript's =npm=, Elixir's =mix=, Go's toolchain, or even the original =rebar=, you might find some behaviours confusing. With that being said, let's use dependencies.

Reviewer: Should mention somewhere, maybe in that list of properties above this, that there is plenty of inspiration of other tools, like `_checkouts` are from leiningen, the dependency resolution is from Maven, being declarative is influenced by both rebar2 and Cargo, and the "pretty" output is based on mix and Cargo I think.

Author: Yeah, that might be good, but I haven't really found a spot where it works best.

**** Declaring Dependencies

> Since dependencies are all project-local, they have to be declared in the =rebar.config= file of your project. This will let Rebar3 know it needs to fetch them, build them, and make them available to your project. All dependencies must be individual OTP applications, so that they can be versioned and handled independently from each other.

Author: I think `Rebar3` is likely going to be simplest to follow when not referring to code, as it visually parses as a name.

]}.
#+END_SRC

Additionally, plugins allow to define [[https://www.rebar3.org/docs/custom-dep-resources][custom resource definitions]] that let you add new types of dependencies to projects.

Let's see how this would work with a project we've created specifically for this book, [[https://github.com/adoptingerlang/service_discovery][=service_discovery=]. Open up the =rebar.config= file and you'll see:

#+NAME: service_discovery.deps
#+BEGIN_SRC erlang
...

{deps, [
    {erldns,
     {git, "https://github.com/tsloughter/erldns.git",
     {branch, "revamp"}}},
    {dns,
     {git, "https://github.com/tsloughter/dns_erlang.git",
     {branch, "hex-deps"}}},

    recon,
    eql,
    jsx,
    {uuid, "1.7.5", {pkg, uuid_erl}},
    {elli, "~> 3.2.0"},
    {grpcbox, "~> 0.11.0"},
    {pgo,
     {git, "https://github.com/tsloughter/pgo.git",
     {branch, "master"}}}
]}.

...
#+END_SRC

Both source and hex dependencies can work together for most projects. The only exception is hex packages, which can only depend on other hex packages. Let's compile the whole project and step through what is going on:

> TODO: FILL IN THE STEPS: show providers, build steps

Author: Ah I thought I had already included this content. I'd start by keeping the content as-is to match what rebar3 does, but I'm fine with updating it.

> TODO: FILL IN THE STEPS: show providers, build steps

Author: Actually we might get a duplicated version with a distinct hash skipped I believe?

**** Building a Project with Dependencies

Building actual projects requires more than just fetching its dependencies and compiling them. As mentioned in [[What Makes a Lib an App]], the Erlang run-time system expects to find run-time definitions of dependencies within the =.app= file. Not putting them there means that while your libraries are available on disk, they won't necessarily be visible during some build phases, some tests, or in releases.

For everything to work, we must add them. Open up =apps/service_discovery/src/service_discovery.app.src= and look at the values in the =applications= tuple:

#+NAME: service_discovery.app-src
#+BEGIN_SRC erlang
{application, service_discovery,
 [{description, "Core functionality for service discovery service"},
  {vsn, {git, short}},
  {registered, []},
  {mod, {service_discovery_app, []}},
  {applications,
   [kernel,
    stdlib,
    erldns,
    service_discovery_storage
   ]},
  {env,[]},
  {modules, []},

  {licenses, ["Apache 2.0"]},
  {links, []}
 ]}.
#+END_SRC

> You can see =erldns= and =service_discovery_storage= have been added. Only when such dependencies are specified will everything both build and run fine.

Author: They may not build fine since I think we currently rely on these tuples for the `-include_lib(...)` directives, breaking the distinction between buildtime and runtime.

In such cases you'd want to have a project configuration that looks like this part of =service_discovery=:

#+NAME: service_discovery.relx-cfg
#+BEGIN_SRC erlang
{relx, [
    {release, {service_discovery, {git, long}},
     [service_discovery_postgres,
      service_discovery,
      service_discovery_http,
      service_discovery_grpc,
      recon]},
    ...
]}.
#+END_SRC

> You can see that, along with our applications, the =recon= debugging tool is explicitly included in the release's list of applications. All their transitive dependencies will be included (according to the =applications= tuple in the =.app= file), but the various OTP applications are handled in a disjoint manner.

Author: I don't know. In the previous review you said that this stuff was generally supported so I didn't question it.

> You can see that, along with our applications, the =recon= debugging tool is explicitly included in the release's list of applications. All their transitive dependencies will be included (according to the =applications= tuple in the =.app= file), but the various OTP applications are handled in a disjoint manner.

Reviewer: It probably does, I just always forget how they decide what goes in an application. I know how you'd set it to not go in the app with like `only: :dev` but not sure how that works for if you don't want it as only a dev dep but only included in the release, and not as an application's dep. I'll look.

Let's focus on these 4 =service_discovery= applications just a bit. These represent the second type of situation where we want to split the declaration of dependencies for builds in =rebar.config=, and the declaration of dependencies for runtime in =.app.src= files.

You can see that at the top-level, all the dependencies for all the libraries are declared in a single =rebar.config= file. That makes it easy for developers to handle and update all versions required. However, if you go look into the =.app.src= files of both =service_discovery= and =service_discovery_http=, you will find this:

#+NAME: service_discovery.app-src-compare
#+BEGIN_SRC erlang
%% service_discovery.app.src
...
  {applications,
   [kernel,
    stdlib,
    erldns,
    service_discovery_storage
   ]},
...
%% service_discovery_http.app.src
...
  {applications,
   [kernel,
    stdlib,
    service_discovery,
    jsx,
    elli
   ]},
...
#+END_SRC

> Here, =service_discovery_http= depends on a web server (=elli=), but =service_discovery= does not. This allows for cleaner boot and shutdown scenarios, where you don't actually need the HTTP server to be up and running to actually start booting the back-end of the system.

Author: Yeah I believe I have a thing somewhere about "having to use a config file per app with duplicated content" or something in rebar3, but I don't bridge it with Mix. Mostly I think documenting what mix does is probably out of scope here.

For a third scenario, you might also imagine a small app =service_discovery_mgmt= that is used only to generate an escript that lets you do system administration tasks to interact with the system and send commands around.

> One option would be to re-declare all its compile-time dependencies in an application-specific =rebar.config= file, but that would be cumbersome; =rebar3= allows to keep everything in the top-level config file. However, if the run-time dependencies were shared across all applications with that top-level declaration, then even if =service_discovery_mgmt= were to _not_ be included in the release (it'd just be a script on the side), its dependencies would still risk being pushed to production in it through other apps getting them automatically inserted in there. Possibly worse, all the dependencies of the =service_discovery= release would risk being bundled with the script as well! We could end up with a small admin tool that contains web servers and database drivers because the build tool tried to be nice.

Author: I think if we are to document mix instead, which the reader never uses, I'd rather take the whole thing out.

> One option would be to re-declare all its compile-time dependencies in an application-specific =rebar.config= file, but that would be cumbersome; =rebar3= allows to keep everything in the top-level config file. However, if the run-time dependencies were shared across all applications with that top-level declaration, then even if =service_discovery_mgmt= were to _not_ be included in the release (it'd just be a script on the side), its dependencies would still risk being pushed to production in it through other apps getting them automatically inserted in there. Possibly worse, all the dependencies of the =service_discovery= release would risk being bundled with the script as well! We could end up with a small admin tool that contains web servers and database drivers because the build tool tried to be nice.

Reviewer: Yea, probably a good idea to just remove.

The rebar3 maintainers therefore just decided to keep a clear distinction between the libraries that need fetching for the project (in =rebar.config=) and the run-time dependencies of OTP applications (in the =.app= file). Other build tools in the ecosystem let you achieve similar results, but they default to including everything at run-time whereas rebar3 asks of developers to always be specific in their intent.

With this all in place, all we have to do is groom and clean our sets of dependencies.

*** Dependency Lifecycles

> When it comes to setting up a project, the dependency resolution and fetching is where the heaviest work is done. The result is stored in a lock file, and all the results after that are handled through shorter partial changes done through rebar3. This first phase is important to understand, even if you need to do it rather infrequently.

Author: I'll rework as well. "a project build environment" to me sounds like setting up the CI and whatnot.

The rebar3 lock file is created at the root of your project, the directory from which you called rebar3. It is saved as =rebar.lock= and you should track it in your source control of choice. You can open the file and look at its content, but there shouldn't be any need to ever edit it by hand. You'll find that it mostly contains version numbers, application names, and various hashes. It might be interesting to audit it from time to time, but you'll end up doing that indirectly as you maintain your dependency tree.

The lock file represents the flattened tree of all dependencies as desired at the time of the build. It will not be modified unless you ask for it to be changed, or unless you flush it and force a new re-resolution to be done from scratch. This is on purpose, and is part of how rebar3 can guarantee repeatable builds under all circumstances.

> The hashes in the file mean that even if the dependencies are fetched by multiple layered mirrors (TODO: add link to section), and some malevolent person alters the packages in one or more of the various hex indices or git sources you might use, rebar3 will be able to find that the information is not as expected, and warn you about it.

Author: I think it will yeah. Will change.

- =rebar3 deps= will list out dependencies and annotate those that could be updated. Do note that there are strong limitations in what it considers worth updating: a branch in git, a tag whose reference shifted, or hex versions that weren't specified. But if you specified a package to have version ="1.2.3"= and =1.2.4= is available, it won't tell you anything

There are other commands that are a bit more drastic, namely =rebar3 unlock= and =rebar3 upgrade= (without any arguments). These will just get rid of the lock file for the next build of the project. But overall, all these commands will do everything you need to manage most dependencies.

In general your workflow might look like this:

1. set up the initial project, compile once, and track the lock file in source control
2. find out you have a top-level dependency you want to change
3. change the dependency definition in the rebar.config file (or optionally if using a non-versioned hex package, call =rebar3 update=)
4. call =rebar3 upgrade <app>= to update the application and its transitive dependencies

And that's it, you're done.

*** Checkout Dependencies

> 

Reviewer: ```suggestion

Rebar3 wants to make the developers life easier while also staying safe and repeatable. =_checkouts= are a feature for keeping a quality of life task that goes against the lock file and repeatability, quick feedback look on local changes to dependencies, simple and controlled.

```

Good to introduce that the feature is for local dependencies quick I think, and to setup for the fact that it will result in the dep being removed from the lock file. I don't think this intro is done but is roughly what I had in mind.

> If you're just adopting Erlang, whether for fun or at work, chances are you'll have one project to maintain, maybe a few more than that, and small ones at that. Working with a few repositories won't be too hard. But sooner or later, if you have to start patching dependencies or if you're working in a corporate environment with dozens and dozens of repositories, working with rebar3 might become frustrating.

Reviewer: "maybe a few more than that, and small ones at that" -- Needs to be reworded. Maybe just say you'll have projects with dependencies that are separate repos?

> The problem would be that every time you want to try and modify a dependency to do something that its parent application needs, you'll have to commit and publish the change somewhere rebar3 can fetch it, because it will otherwise keep trying to make a repeatable build by fetching remote resources without your fix, or won't recompile things it believes are up to date. This can become annoying really fast when you have to work across repository boundaries and just want to test a change.

Reviewer: ```suggestion
The problem would be that every time you want to try a modification of a dependency you will have to commit and publish the change somewhere rebar3 can fetch it, because it will not build changes made to source files under `_build` if there is already a corresponding `.beam` file for the module. This can become annoying really fast when you have to work across repository boundaries and just want to test a change.
```

I try to hint at the fact that they can do what I do sometimes and just `rm _build/default/<dep>/ebin/<module>.beam` to pick up changes to `_build/default/<dep>/src/<module>.erl` but without actually suggesting that :)

Author: We might actually want to add a switch like `rebar3 clean -d <appname>` to only flush that one app.

Reviewer: Yea, that would be good.

So there's a little trickshot of a feature called _checkout dependencies_. Checkout dependencies work as follows:

- you have your main project somewhere on your file system
- the dependency is declared in the rebar.config file of the main project
- you also have that dependency somewhere on your file system, as a standalone project
- you add a =_checkouts/= directory to the main project
- you either copy or symlink the dependency's directory in the =_checkouts/= directory

> From that point on, every time the main app is built, it will add an =ebin/= directory to the dependency's directory in checkouts, and re-compile it as if it were a top-level application within the main app.

Reviewer: ```suggestion
From that point on, every time the main app is built, it will add an =ebin/= directory to the dependency's directory in checkouts, and re-compile it as if it were a top-level application within the main project.
```

> You can then test the changes to your dependency within the main application until they're ready. Once you're done, remove the =ebin/= directory from the dependency, commit and publish your code in the dependency, remove it from the =_checkouts= directory, and =rebar3 upgrade= it.

Reviewer: ```suggestion
You can then test the changes to your dependency within the main project until they're ready. Once you're done, remove the =ebin/= directory from the dependency, commit and publish your code in the dependency, remove it from the =_checkouts= directory, and =rebar3 upgrade= it.
```

> This lets you do a lot of small iterative changes locally on a dependency, within the context of the main app, without having to make everything public all the time nor changing configuration files to point a dep to some local directory. It massively reduces the overhead of per-app repositories as an overall development strategy.

Reviewer: ```suggestion
This lets you do a lot of small iterative changes locally on a dependency, within the context of the main project, without having to push the dependency's changes nor changing configuration files to point a dependency to some local directory. It massively reduces the overhead of per-application repositories as an overall development strategy.
```

With Elixir in place, add the [[https://www.rebar3.org/docs/using-available-plugins#section-elixir-dependencies][rebar_mix]] plugin to your library or project:

#+NAME: rebar_mix.config
#+BEGIN_SRC erlang
{plugins, [rebar_mix]}.
{provider_hooks, [{post, [{compile, {mix, consolidate_protocols}}]}]}.

{relx, [
    ...
    {overlay, [
        {copy, "{{base_dir}}/consolidated", "releases/{{release_version}}/consolidated"}
    ]
}.
#+END_SRC

And the following line to your =vm.args.src= file if you have one:

#+NAME: rebar_mix.args
#+BEGIN_SRC sh
-pa releases/${REL_VSN}/consolidated
#+END_SRC

From that point on, you will be able to install any hex dependency containing Elixir code without a problem. For now, the plugin only supports hex dependencies that also only rely on other hex dependencies; support for transitive dependencies using git should be coming soon, however.

Do note that mixed rebar3 projects using both Erlang and Elixir within the same library are not currently supported since more work both on the rebar3 and Elixir side would need to be done to make this possible. On the other hand, Mix does support this pattern if you need it.

*** Corporate Environments

> Corporate environments tend to have all kinds of weird restrictions regarding what can or cannot be done, and development tools can be very idiosyncratic. Rebar3 was mostly developed to fit an open-sourced world and focuses mostly on enforcing project structure, fetching dependencies, and using both together as a great pretext to wrap a bunch of standard tools.

Reviewer: ```suggestion
Corporate environments tend to have all kinds of weird restrictions regarding what can or cannot be done, and development tools can be very idiosyncratic. Rebar3 was mostly developed to fit an open-sourced world and mainly focuses on enforcing project structure, fetching dependencies, and using both together as a great pretext to wrap a bunch of standard tools.
```

As such, it is perhaps unsurprising that fitting a corporate environment with the tool can prove to be a bit challenging. In this section, we'll cover some standard tools that are common to corporate environments and may make your life easier when adopting Erlang.

**** Proxy Support

> Many workplaces enforce very strict firewall rules, to the point where all incoming and outgoing data must be intercepted and monitored. Generally, these places will not be totally isolated from the public Internet, but will require the use of proxy servers to make all connections work.

Reviewer: ```suggestion
Many workplaces enforce very strict firewall rules, to the point where all incoming and outgoing data must be intercepted and monitored. Generally, these places will not be totally isolated from the public Internet, but will require the use of proxy servers to make outgoing connections.
```

> It is rather standard for open source program to respect both the =HTTP_PROXY= and =HTTPS_PROXY= environment variable. When those are set in the development environment you use, rebar3 will make sure that all the communications it makes talking to the outside world will reuse these proxies.

Reviewer: ```suggestion
It is rather standard for programs to respect both the =HTTP_PROXY= and =HTTPS_PROXY= environment variable. When those are set in your development environment rebar3 will make sure that all the communications it makes talking to the outside world uses these proxies.
```

Rebar3 supports this use case. If you want to enable it, you will need to first set up a private hex instance. This can be done through the [[http://blog.plataformatec.com.br/2019/07/announcing-minirepo-a-minimal-hex-server/][minirepo]] project. By following the instructions [[https://github.com/wojtekmach/mini_repo][on the project page]], you will end up running your own private hex server, with either a local filesystem or S3-based storage, and the ability to both mirror other indexes and publish your own packages privately.

You will need to [[https://github.com/wojtekmach/mini_repo#usage-with-rebar3][tweak your global rebar3 config]] to make use of it, but once that's done, you're good to go.

**** About Monorepos

> If you work in a corporate environment with a monorepo where all private libraries, OTP applications, and dependencies are treated on equal footing (it's not just an umbrella release), rebar3 is not the best tool for the job. Mostly, this comes down to the fact that most companies using a monorepo have a lot of custom tooling with very custom workflows, large codebases, and a very strong propension to never share access with maintainers of rebar3. Until it becomes possible for maintainers to get access, little will be possible to do on that front.

Reviewer: Maybe mention benoit's vendor plugin?

> Various commercial users who rely on rebar3 despite using monorepos have reported obtaining successful builds with a combination of using =_checkout= dependencies, along with re-configuring the =_build= directory. However, we cannot recommend this approach at this time, and no official support is provided for monorepos.

Reviewer: ```suggestion
Various commercial users who rely on rebar3 despite using monorepos have reported obtaining successful builds with a combination of using =_checkouts= dependencies, along with re-configuring the =_build= directory. However, we cannot recommend this approach at this time, and no official support is provided for monorepos.
```

** External Services
:PROPERTIES:
:EXPORT_FILE_NAME: external_services
:EXPORT_HUGO_MENU: :menu main :parent development
:END:

> 

Author: Oh yeah. It's probably a merge that went difficult during the rebase.
